File: dashboard/aux_claude_usage.py
# aux_claude_usage.py — "Claude Usage" hub widget (local-first, read-only).
#
# exec'd into server.py's globals by the aux-module loader (after
# expanders_extra.py, sorted before the other aux_*.py). It may use these
# server.py globals: HOME, read_json, write_json, _cached, _widget_cache,
# register_get, register_post, get_layout, save_layout, WIDGETS, EXPANDERS.
# It imports ALL its own stdlib deps (exec'd code can't rely on server.py's
# function-local imports) and defines only new names (CU_*, _cu_*,
# w_claude_usage, expand_claude_usage) so it clobbers nothing.
#
# What it does: scans Claude Code's session logs at ~/.claude/projects/**/*.jsonl
# (assistant records carry message.usage + message.model + a top-level ISO
# timestamp), and reports the rolling 5-hour rate-limit window, today, a 7-day
# series, per-model / per-project breakdowns, block history, and an
# "≈ API-equivalent" cost. Read-only on ~/.claude; efficient (60s cache, bounded
# scan, skips files older than the window); never raises the hub down.
#
# Rate-limit modeling: Claude Max resets on a ROLLING 5-hour window (plus weekly
# caps) and publishes no hard token number, so we don't fabricate one. Instead we
# reconstruct the 5-hour "blocks" the way ccusage does — a block starts at the
# first activity (floored to the hour) and spans 5h; a gap >5h opens a new block —
# and surface the ACTIVE block's usage + its reset countdown, plus today/7-day
# totals and a sparkline against the user's own recent peak. An OPTIONAL soft cap
# (stored in our own JSON, no server.py edit) turns that into a % gauge.

import os
import sys
import json
import time
import datetime as _cu_datetime   # private alias: other aux modules do
                                  # `from datetime import datetime`, which would
                                  # rebind a bare `datetime` global to the class.

# --------------------------------------------------------------------------
# constants
# --------------------------------------------------------------------------
CU_PROJECTS    = os.path.join(HOME, ".claude", "projects")
CU_CONFIG_FILE = os.path.join(HOME, ".hermes", "dashboard", "claude-usage.json")

CU_BLOCK_SECS = 5 * 3600      # the rolling 5-hour rate-limit window
CU_SCAN_DAYS  = 8             # only read files modified within this many days
CU_MAX_FILES  = 800           # bound the scan
CU_MAX_LINES  = 400000        # per-file safety cap

# Published Claude API prices ($/1M tok in, $/1M tok out), keyed by model-family
# substring. Cost is labelled "≈ API-equivalent" — Max is a flat subscription.
# cache_read ~= 0.1x input, cache_creation (write) ~= 1.25x input.
CU_PRICES = {
    "fable":  (10.0, 50.0),
    "mythos": (10.0, 50.0),
    "opus":   (5.0, 25.0),
    "sonnet": (3.0, 15.0),
    "haiku":  (1.0, 5.0),
}
CU_DEFAULT_PRICE = (3.0, 15.0)   # unknown family -> Sonnet-tier (neutral middle)
CU_CACHE_READ_MULT = 0.1
CU_CACHE_WRITE_MULT = 1.25


# --------------------------------------------------------------------------
# helpers
# --------------------------------------------------------------------------
def _cu_family(model):
    m = (model or "").lower()
    for fam in ("fable", "mythos", "haiku", "sonnet", "opus"):
        if fam in m:
            return fam
    return "other"


def _cu_price(model):
    return CU_PRICES.get(_cu_family(model), CU_DEFAULT_PRICE)


def _cu_cost(model, inp, out, cr, cc):
    pin, pout = _cu_price(model)
    return (inp * pin + cr * pin * CU_CACHE_READ_MULT +
            cc * pin * CU_CACHE_WRITE_MULT + out * pout) / 1e6


def _cu_ts(rec):
    """Top-level ISO timestamp (e.g. 2026-07-05T15:27:00.439Z) -> epoch secs."""
    t = rec.get("timestamp")
    if not isinstance(t, str):
        return None
    try:
        return _cu_datetime.datetime.fromisoformat(t.replace("Z", "+00:00")).timestamp()
    except Exception:
        return None


def _cu_cfg():
    c = read_json(CU_CONFIG_FILE, {})
    return c if isinstance(c, dict) else {}


# --------------------------------------------------------------------------
# scan: walk ~/.claude/projects/**/*.jsonl (read-only, bounded, fault-tolerant)
# --------------------------------------------------------------------------
def _cu_scan():
    events = []            # (ts, family, model, in, out, cache_read, cache_write)
    sessions = set()
    proj_tok = {}          # project-dir -> aggregate
    proj_cwd = {}          # project-dir -> real cwd (for a friendly name)
    now = time.time()
    cutoff = now - CU_SCAN_DAYS * 86400

    files = []
    try:
        for root, dirs, fnames in os.walk(CU_PROJECTS):
            for fn in fnames:
                if not fn.endswith(".jsonl"):
                    continue
                p = os.path.join(root, fn)
                try:
                    st = os.stat(p)
                except OSError:
                    continue
                if st.st_mtime < cutoff:
                    continue
                files.append((st.st_mtime, p))
    except Exception:
        pass
    files.sort(reverse=True)           # newest first
    files = files[:CU_MAX_FILES]

    for _mt, p in files:
        try:
            rel = os.path.relpath(os.path.dirname(p), CU_PROJECTS)
            proj = rel.split(os.sep)[0] if rel and rel != "." else "?"
        except Exception:
            proj = "?"
        try:
            with open(p, "r", encoding="utf-8", errors="replace") as f:
                n = 0
                for line in f:
                    n += 1
                    if n > CU_MAX_LINES:
                        break
                    if '"usage"' not in line:      # cheap skip before json.loads
                        continue
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        rec = json.loads(line)
                    except Exception:
                        continue
                    if not isinstance(rec, dict):
                        continue
                    msg = rec.get("message")
                    if not isinstance(msg, dict):
                        continue
                    usage = msg.get("usage")
                    if not isinstance(usage, dict):
                        continue
                    ts = _cu_ts(rec)
                    if ts is None:
                        continue
                    model = msg.get("model") or "?"
                    try:
                        inp = int(usage.get("input_tokens") or 0)
                        out = int(usage.get("output_tokens") or 0)
                        cr = int(usage.get("cache_read_input_tokens") or 0)
                        cc = int(usage.get("cache_creation_input_tokens") or 0)
                    except (TypeError, ValueError):
                        continue
                    if (inp + out + cr + cc) <= 0:
                        continue
                    events.append((ts, _cu_family(model), model, inp, out, cr, cc))
                    sid = rec.get("sessionId")
                    if sid:
                        sessions.add(sid)
                    if proj not in proj_cwd:
                        cw = rec.get("cwd")
                        if isinstance(cw, str) and cw:
                            proj_cwd[proj] = cw
                    pa = proj_tok.get(proj)
                    if pa is None:
                        pa = proj_tok[proj] = {"in": 0, "out": 0, "cr": 0,
                                               "cc": 0, "cost": 0.0, "msgs": 0}
                    pa["in"] += inp; pa["out"] += out
                    pa["cr"] += cr; pa["cc"] += cc
                    pa["cost"] += _cu_cost(model, inp, out, cr, cc)
                    pa["msgs"] += 1
        except OSError:
            continue

    events.sort(key=lambda e: e[0])
    return events, sessions, proj_tok, proj_cwd


# --------------------------------------------------------------------------
# build: aggregate the scanned events into the full payload
# --------------------------------------------------------------------------
def _cu_agg():
    return {"in": 0, "out": 0, "cr": 0, "cc": 0, "total": 0, "cost": 0.0, "msgs": 0}


def _cu_add(a, e):
    _ts, _fam, model, inp, out, cr, cc = e
    a["in"] += inp; a["out"] += out; a["cr"] += cr; a["cc"] += cc
    # "total" is REAL work = input + output. Cache-READ tokens (the same cached
    # prompt prefix re-served on every message) are NOT counted here — including
    # them made the headline 10-100x too high (a session's context is re-read
    # every turn). Cache read/write are surfaced separately as `cache`.
    a["total"] += inp + out
    a["cost"] += _cu_cost(model, inp, out, cr, cc)
    a["msgs"] += 1


def _cu_pack(a):
    return {"in": a["in"], "out": a["out"], "cr": a["cr"], "cc": a["cc"],
            "cache": a["cr"] + a["cc"], "total": a["total"],
            "cost": round(a["cost"], 4), "msgs": a["msgs"]}


def _cu_proj_name(name, cwd):
    if isinstance(cwd, str) and cwd:
        base = os.path.basename(cwd.rstrip("/"))
        return base or cwd
    return (name or "?").lstrip("-")[:40] or "?"


def _cu_build():
    events, sessions, proj_tok, proj_cwd = _cu_scan()
    now = time.time()

    # --- rolling 5-hour blocks (ccusage-style) ---
    blocks = []
    cur = None
    last_ts = None
    for e in events:
        ts = e[0]
        new_block = (cur is None
                     or (ts - cur["start"] >= CU_BLOCK_SECS)
                     or (last_ts is not None and ts - last_ts >= CU_BLOCK_SECS))
        if new_block:
            start = ts - (ts % 3600)          # floor first activity to the hour
            cur = _cu_agg()
            cur["start"] = start
            cur["end"] = ts
            blocks.append(cur)
        cur["end"] = ts
        _cu_add(cur, e)
        last_ts = ts

    active = None
    if blocks:
        b = blocks[-1]
        if now < b["start"] + CU_BLOCK_SECS:  # still inside its 5-hour window
            active = b

    # --- local day boundaries ---
    lt = time.localtime(now)
    midnight = time.mktime((lt.tm_year, lt.tm_mon, lt.tm_mday, 0, 0, 0, 0, 0, -1))
    wk_cut = midnight - 6 * 86400

    today = _cu_agg()
    week = _cu_agg()
    for e in events:
        if e[0] >= midnight:
            _cu_add(today, e)
        if e[0] >= wk_cut:
            _cu_add(week, e)

    # --- 7-day daily series (oldest -> today) ---
    days = []
    day_aggs = [_cu_agg() for _ in range(7)]
    for e in events:
        if e[0] < wk_cut:
            continue
        idx = int((e[0] - wk_cut) // 86400)
        if 0 <= idx <= 6:
            _cu_add(day_aggs[idx], e)
    for i in range(7):
        d0 = wk_cut + i * 86400
        a = day_aggs[i]
        days.append({"date": time.strftime("%Y-%m-%d", time.localtime(d0 + 3600)),
                     "total": a["total"], "in": a["in"], "out": a["out"],
                     "cr": a["cr"], "cc": a["cc"], "cost": round(a["cost"], 4),
                     "msgs": a["msgs"]})

    # --- per-model (7 days) ---
    models = {}
    for e in events:
        if e[0] < wk_cut:
            continue
        fam = e[1]
        ma = models.get(fam)
        if ma is None:
            ma = models[fam] = _cu_agg()
            ma["family"] = fam
        _cu_add(ma, e)
    model_list = []
    for ma in sorted(models.values(), key=lambda m: -m["total"]):
        row = _cu_pack(ma)
        row["family"] = ma["family"]
        model_list.append(row)

    # --- per-project ---
    projs = []
    for name, pa in proj_tok.items():
        total = pa["in"] + pa["out"]          # real work (cache-read excluded)
        projs.append({"name": _cu_proj_name(name, proj_cwd.get(name)),
                      "path": proj_cwd.get(name), "total": total,
                      "cache": pa["cr"] + pa["cc"],
                      "cost": round(pa["cost"], 4), "msgs": pa["msgs"]})
    projs.sort(key=lambda p: -p["total"])

    # --- block history (recent) ---
    bhist = []
    for b in blocks[-14:]:
        row = _cu_pack(b)
        row["start"] = b["start"]
        row["end"] = b["end"]
        row["active"] = (active is not None and b["start"] == active["start"])
        bhist.append(row)

    day_peak = max((d["total"] for d in days), default=0)
    block_peak = max((b["total"] for b in blocks), default=0)

    # --- current 5-hour window ---
    if active is not None:
        window = _cu_pack(active)
        window["active"] = True
        window["start"] = active["start"]
        window["reset_at"] = active["start"] + CU_BLOCK_SECS
        window["reset_in"] = max(0, active["start"] + CU_BLOCK_SECS - now)
    else:
        window = _cu_pack(_cu_agg())
        window["active"] = False
        window["start"] = None
        window["reset_at"] = None
        window["reset_in"] = 0

    cfg = _cu_cfg()
    return {
        "available": True,
        "generated": now,
        "block_hours": 5,
        "scan_days": CU_SCAN_DAYS,
        "window": window,
        "today": _cu_pack(today),
        "week": _cu_pack(week),
        "days": days,
        "day_peak": day_peak,
        "block_peak": block_peak,
        "models": model_list,
        "projects": projs[:12],
        "blocks": bhist,
        "sessions": len(sessions),
        "messages": len(events),
        "cap": cfg.get("cap"),
        "plan": cfg.get("plan"),
        "prices": {k: {"in": v[0], "out": v[1]} for k, v in CU_PRICES.items()},
    }


def _cu_data():
    return _cached("claude_usage", 60, _cu_build)


# --------------------------------------------------------------------------
# widget provider (compact body data) + rich expander
# --------------------------------------------------------------------------
def w_claude_usage():
    try:
        d = _cu_data()
    except Exception as e:
        return {"available": False, "reason": type(e).__name__}
    return {
        "available": d.get("available", True),
        "window": d["window"],
        "today": d["today"],
        "week_total": d["week"]["total"],
        "spark": [x["total"] for x in d["days"]],
        "day_peak": d["day_peak"],
        "block_peak": d["block_peak"],
        "cap": d.get("cap"),
        "block_hours": d["block_hours"],
        "sessions": d["sessions"],
        "messages": d["messages"],
        "generated": d["generated"],
    }


def expand_claude_usage():
    try:
        return _cu_data()
    except Exception as e:
        return {"available": False, "reason": "%s: %s" % (type(e).__name__, e)}


# --------------------------------------------------------------------------
# config route: optional soft cap / plan tier (own JSON, no server.py edit)
# --------------------------------------------------------------------------
def _cu_config_get(ctx):
    return {"ok": True, "config": _cu_cfg()}


def _cu_config_post(ctx):
    b = ctx.body or {}
    cfg = _cu_cfg()
    if "cap" in b:
        cap = b.get("cap")
        if cap in (None, "", 0, "0"):
            cfg.pop("cap", None)
        else:
            try:
                cfg["cap"] = max(0, int(cap))
            except (TypeError, ValueError):
                return ({"ok": False, "error": "bad_cap"}, 400)
    if "plan" in b:
        plan = b.get("plan")
        if plan:
            cfg["plan"] = str(plan)[:40]
        else:
            cfg.pop("plan", None)
    try:
        os.makedirs(os.path.dirname(CU_CONFIG_FILE), exist_ok=True)
        write_json(CU_CONFIG_FILE, cfg)
    except Exception as e:
        return ({"ok": False, "error": "internal: " + str(e)}, 500)
    try:
        _widget_cache.pop("claude_usage", None)   # reflect the cap immediately
    except Exception:
        pass
    return {"ok": True, "config": cfg}


# --------------------------------------------------------------------------
# module-load side effects: register routes, catalog entry, layout injection
# --------------------------------------------------------------------------
register_get("/api/claude_usage/config", _cu_config_get)
register_post("/api/claude_usage/config", _cu_config_post)

WIDGETS["claude_usage"] = {"title": "Claude Usage", "icon": "activity",
                           "size": "card", "cat": "agent",
                           "provider": w_claude_usage}
EXPANDERS["claude_usage"] = expand_claude_usage

# Show up without a manual add: append to the layout order IF absent (never
# clobber the user's order). WIDGETS already has claude_usage above, so
# get_layout()'s catalog filter keeps it.
try:
    _cu_lay = get_layout()
    if isinstance(_cu_lay, dict):
        _cu_order = _cu_lay.get("order")
        if not isinstance(_cu_order, list):
            _cu_order = _cu_lay["order"] = []
        if "claude_usage" not in _cu_order:
            _cu_order.append("claude_usage")
            save_layout(_cu_lay)
except Exception as _cu_e:                                   # pragma: no cover
    print("[aux_claude_usage] layout inject failed: %s" % _cu_e, file=sys.stderr)

File: dashboard/aux_claude_usage.js
// aux_claude_usage.js — "Claude Usage" widget renderers (P: Claude Max tracker).
//
// Auto-served at /aux_claude_usage.js. Loaded AFTER /expand.js and the other
// aux JS, so these map assignments win. Adds:
//   * WICONS.claude_usage       — bespoke two-tone gauge glyph (widgetIcon)
//   * RENDER.claude_usage       — compact widget body (current 5h window +
//                                 reset countdown + today + 7-day sparkline)
//   * EXPAND_RENDER.claude_usage — pop-out (5h block detail + reset, today/7-day
//                                 chart, per-model, per-project, block history,
//                                 ≈ API-equivalent cost, soft-cap setter).
//
// Reuses global helpers from index.html (esc, kfmt, fmtNum, statGrid, barRows,
// animate) — all typeof-guarded so a headless render harness never throws.
// CLAUDE.md laws: zero emoji, bespoke SVG only, 12-hour time, esc() coercion.

(function () {
  "use strict";

  // ---- tiny helpers (guarded so headless eval can't throw) -----------------
  function E(s) { return (typeof esc === "function") ? esc(s) : String(s == null ? "" : s); }
  function KF(n) {
    if (typeof kfmt === "function") return kfmt(n || 0);
    n = +n || 0;
    if (n >= 1e6) return (n / 1e6).toFixed(1) + "M";
    if (n >= 1e3) return (n / 1e3).toFixed(1) + "k";
    return "" + n;
  }
  function NF(n) {
    if (typeof fmtNum === "function") return fmtNum(n == null ? 0 : n);
    return (+n || 0).toLocaleString();
  }
  function money(n) {
    n = +n;
    if (!isFinite(n)) n = 0;
    if (n >= 100) return "$" + Math.round(n).toLocaleString();
    return "$" + n.toFixed(2);
  }
  function t12(sec) {           // epoch seconds -> "3:45 PM" (12-hour)
    if (sec == null) return "";
    try {
      return new Date(sec * 1000).toLocaleTimeString([], { hour: "numeric", minute: "2-digit" });
    } catch (e) { return ""; }
  }
  function dur(sec) {           // seconds -> "2h 10m"
    sec = Math.max(0, Math.round(sec || 0));
    var h = Math.floor(sec / 3600), m = Math.floor((sec % 3600) / 60);
    if (h > 0) return h + "h " + m + "m";
    if (m > 0) return m + "m";
    return sec + "s";
  }
  function sub(t) {
    return '<div style="font-size:10px;text-transform:uppercase;letter-spacing:.06em;' +
      'color:var(--muted);font-weight:600;margin:12px 0 6px">' + E(t) + "</div>";
  }
  function gaugeBar(pct, label, color) {
    pct = Math.max(2, Math.min(100, pct || 0));
    return '<div style="height:8px;border-radius:5px;background:var(--hairline);overflow:hidden;margin:2px 0 5px">' +
      '<i style="display:block;height:100%;width:' + pct + '%;border-radius:5px;background:' + color + '"></i></div>' +
      '<div class="w-sub" style="font-size:10.5px;margin-bottom:10px">' + E(label) + "</div>";
  }
  function famName(f) {
    f = String(f || "");
    return f ? (f.charAt(0).toUpperCase() + f.slice(1)) : "Other";
  }
  // full-width 7-day bar sparkline (bespoke SVG; last bar = accent)
  function cuBars(arr, peak, accent) {
    if (!arr || !arr.length) return "";
    var mx = peak || Math.max.apply(null, arr.concat([1])) || 1;
    var n = arr.length, gap = 2.5, bw = (100 - (n - 1) * gap) / n, bars = "";
    for (var i = 0; i < n; i++) {
      var pv = mx > 0 ? (arr[i] / mx) : 0;
      var hh = Math.max(3, pv * 27), last = i === n - 1;
      var col = last ? accent : "color-mix(in srgb," + accent + " 40%,transparent)";
      bars += '<rect x="' + (i * (bw + gap)).toFixed(2) + '" y="' + (29 - hh).toFixed(2) +
        '" width="' + bw.toFixed(2) + '" height="' + hh.toFixed(2) + '" rx="1.3" fill="' + col + '"/>';
    }
    return '<svg viewBox="0 0 100 30" preserveAspectRatio="none" style="width:100%;height:26px;display:block">' +
      bars + "</svg>";
  }

  // ---- bespoke widget icon: a usage gauge -----------------------------------
  if (typeof WICONS !== "undefined") {
    WICONS.claude_usage =
      '<path d="M3.5 17.5a8.5 8.5 0 0 1 17 0Z" fill="currentColor" opacity=".15"/>' +
      '<path d="M3.5 17.5a8.5 8.5 0 0 1 17 0" stroke-linecap="round"/>' +
      '<path d="M12 17.5l4.4-5.6" stroke-linecap="round"/>' +
      '<circle cx="12" cy="17.5" r="1.5" fill="currentColor"/>';
  }

  // ---- compact widget body --------------------------------------------------
  if (typeof RENDER !== "undefined") {
    RENDER.claude_usage = function (body, data, mslot) {
      var d = data || {};
      if (d.available === false) {
        body.innerHTML = '<div class="hint">' + E(d.reason || "No Claude Code logs found yet.") + "</div>";
        return;
      }
      var w = d.window || {}, t = d.today || {};
      var spark = d.spark || [], peak = d.day_peak || 0, accent = "var(--wac)";

      var resetLine;
      if (w.active && w.reset_at) {
        resetLine = "resets " + t12(w.reset_at) + " · " + dur(w.reset_at - Date.now() / 1000) + " left";
      } else {
        resetLine = "window clear · a new prompt starts one";
      }
      if (mslot) mslot.textContent = (d.block_hours || 5) + "h window";

      var capPct = null;
      if (d.cap && d.cap > 0 && w.total != null) capPct = Math.min(100, Math.round(w.total / d.cap * 100));

      var h = "";
      h += '<div style="display:flex;align-items:baseline;gap:8px;margin:2px 0 1px">' +
        '<span class="num" style="font-size:26px;font-weight:720;line-height:1;color:' + accent + '">' + KF(w.total || 0) + "</span>" +
        '<span class="w-sub" style="font-size:11px">tok this 5h</span>' +
        '<span class="num" style="margin-left:auto;font-weight:640;font-size:12.5px">' + money(w.cost) + "</span></div>";
      h += '<div class="w-sub" style="font-size:11px;display:flex;align-items:center;gap:6px;margin-bottom:8px">' +
        (w.active ? '<span class="livedot"></span>' : "") + E(resetLine) +
        (w.cache ? '<span style="margin-left:auto;opacity:.7" title="Cached context re-served each message — not new work, mostly free">+' + KF(w.cache) + " cache</span>" : "") + "</div>";

      if (capPct != null) {
        var cc = capPct >= 90 ? "var(--bad)" : capPct >= 70 ? "var(--warn)" : accent;
        h += '<div style="height:6px;border-radius:4px;background:var(--hairline);overflow:hidden;margin-bottom:4px">' +
          '<i style="display:block;height:100%;width:' + capPct + '%;border-radius:4px;background:' + cc + '"></i></div>' +
          '<div class="w-sub" style="font-size:10.5px;margin-bottom:8px">' + capPct + "% of your " + KF(d.cap) + " soft cap</div>";
      }

      h += '<div style="display:flex;gap:14px;font-size:11px;margin-bottom:6px">' +
        '<span class="w-sub">Today <b class="num" style="color:var(--ink)">' + KF(t.total || 0) + "</b></span>" +
        '<span class="w-sub">7-day <b class="num" style="color:var(--ink)">' + KF(d.week_total || 0) + "</b></span>" +
        '<span class="w-sub" style="margin-left:auto">≈' + money(t.cost || 0) + " today</span></div>";
      h += cuBars(spark, peak, accent);

      body.innerHTML = h;
    };
  }

  // ---- rich pop-out ---------------------------------------------------------
  if (typeof EXPAND_RENDER !== "undefined") {
    EXPAND_RENDER.claude_usage = function (el, d) {
      d = d || {};
      if (d.available === false) {
        el.innerHTML = '<div class="hint">' + E(d.reason || "No Claude Code session logs found.") + "</div>";
        return;
      }
      var w = d.window || {}, t = d.today || {}, wk = d.week || {}, accent = "var(--wac)";
      var h = "";

      // --- current 5-hour window hero ---
      var resetLine;
      if (w.active && w.reset_at) {
        resetLine = "Resets " + t12(w.reset_at) + " · " + dur(w.reset_at - Date.now() / 1000) + " remaining";
      } else {
        resetLine = "No active window — your next prompt opens a fresh 5-hour block.";
      }
      h += '<div style="display:flex;align-items:baseline;gap:10px;margin:2px 0 3px">' +
        '<span class="num" style="font-size:40px;font-weight:730;line-height:1;color:' + accent + '">' + KF(w.total || 0) + "</span>" +
        '<span style="font-size:13px;color:var(--muted)">tokens · 5-hour window</span>' +
        '<span class="num" style="margin-left:auto;font-weight:660;font-size:15px">' + money(w.cost) + "</span></div>";
      h += '<div class="w-sub" style="display:flex;align-items:center;gap:6px;font-size:12px;margin-bottom:10px">' +
        (w.active ? '<span class="livedot"></span>' : "") + E(resetLine) + "</div>";

      // --- rate-limit gauge: soft cap if set, else vs. own busiest block ---
      if (d.cap && d.cap > 0) {
        var cp = Math.min(100, Math.round((w.total || 0) / d.cap * 100));
        var cc = cp >= 90 ? "var(--bad)" : cp >= 70 ? "var(--warn)" : accent;
        h += gaugeBar(cp, KF(w.total || 0) + " / " + KF(d.cap) + " soft cap (" + cp + "%)", cc);
      } else if (d.block_peak) {
        var bp = Math.min(100, Math.round((w.total || 0) / d.block_peak * 100));
        h += gaugeBar(bp, "vs your busiest 5-hour block (" + KF(d.block_peak) + " tok) — no official Max cap published", accent);
      }

      // --- window breakdown ---
      if (typeof statGrid === "function") {
        h += statGrid([
          ["Input", KF(w.in || 0)], ["Output", KF(w.out || 0)],
          ["Cache read", KF(w.cr || 0)], ["Cache write", KF(w.cc || 0)],
          ["Messages", NF(w.msgs || 0)], ["≈ Cost", money(w.cost)],
        ]);
      }

      // --- today vs 7-day ---
      h += sub("Today vs last 7 days");
      if (typeof statGrid === "function") {
        h += statGrid([
          ["Today tok", KF(t.total || 0)], ["Today ≈", money(t.cost)], ["Today msgs", NF(t.msgs || 0)],
          ["7-day tok", KF(wk.total || 0)], ["7-day ≈", money(wk.cost)], ["7-day msgs", NF(wk.msgs || 0)],
        ]);
      }

      // --- 7-day daily bar chart ---
      var days = d.days || [];
      if (days.length) {
        var peak = Math.max.apply(null, days.map(function (x) { return x.total; }).concat([1]));
        h += sub("Daily tokens (7 days)");
        h += '<div style="display:flex;align-items:flex-end;gap:6px;height:92px">';
        days.forEach(function (x, i) {
          var hh = Math.max(3, Math.round((x.total / peak) * 100)), last = i === days.length - 1;
          h += '<div class="cu-bar" title="' + E(x.date + " · " + KF(x.total) + " tok") + '" style="flex:1;height:' + hh +
            "%;border-radius:5px 5px 2px 2px;transform-origin:bottom;background:" +
            (last ? accent : "color-mix(in srgb," + accent + " 42%,transparent)") + '"></div>';
        });
        h += "</div>";
        h += '<div style="display:flex;gap:6px;margin-top:4px">';
        days.forEach(function (x) {
          var dt = new Date((x.date || "") + "T00:00");
          var lab = isNaN(dt.getTime()) ? "" : dt.toLocaleDateString([], { weekday: "short" });
          h += '<div style="flex:1;text-align:center;font-size:9.5px;color:var(--faint)">' + E(lab) + "</div>";
        });
        h += "</div>";
      }

      // --- per-model ---
      var models = d.models || [];
      if (models.length && typeof barRows === "function") {
        h += sub("By model (7 days)");
        h += barRows(models.map(function (m) {
          return { label: famName(m.family), val: m.total, sub: money(m.cost) };
        }), 96, 60);
      }

      // --- per-project ---
      var projs = d.projects || [];
      if (projs.length && typeof barRows === "function") {
        h += sub("By project (7 days)");
        h += barRows(projs.map(function (p) {
          return { label: p.name, val: p.total, sub: money(p.cost) };
        }), 130, 56);
      }

      // --- block history ---
      var blocks = (d.blocks || []).slice().reverse();
      if (blocks.length) {
        h += sub("Recent 5-hour blocks");
        blocks.forEach(function (b) {
          var day = "";
          try { day = new Date(b.start * 1000).toLocaleDateString([], { month: "short", day: "numeric" }); } catch (e) {}
          h += '<div style="display:flex;align-items:center;gap:9px;padding:6px 0;border-bottom:1px solid var(--hairline);font-size:12px">' +
            (b.active ? '<span class="livedot"></span>' : '<span style="width:7px;flex:0 0 7px"></span>') +
            '<span class="num" style="width:126px;color:var(--muted)">' + E(day + " · " + t12(b.start) + "–" + t12(b.end)) + "</span>" +
            '<span class="num" style="font-weight:640;min-width:52px;text-align:right">' + KF(b.total) + "</span>" +
            '<span class="w-sub" style="flex:1;text-align:right;font-size:11px">' + NF(b.msgs) + " msgs · " + money(b.cost) + "</span></div>";
        });
      }

      // --- totals + cost caveat ---
      h += '<div class="w-sub" style="margin:11px 0 4px;font-size:11px">' +
        NF(d.sessions || 0) + " sessions · " + NF(d.messages || 0) +
        " assistant messages in the last " + (d.scan_days || 8) + " days</div>";
      h += '<div class="w-sub" style="font-size:10.5px;color:var(--faint);line-height:1.5;margin-bottom:10px">' +
        "Cost is <b>≈ API-equivalent</b> (published per-token prices) — your Max plan is a flat subscription, " +
        "so this is what the same tokens would bill on the API, not a charge. Max publishes no hard token limit; the " +
        "5-hour window above is the rolling reset Claude enforces.</div>";

      // --- optional soft-cap setter ---
      h += '<div style="display:flex;align-items:center;gap:8px;margin-bottom:4px">' +
        '<input id="cu-cap" placeholder="Soft cap (tokens / 5h)" ' +
        'style="flex:1;padding:7px 10px;border-radius:9px;border:1px solid var(--hairline);background:var(--glass-2);color:var(--ink);font:inherit" ' +
        'value="' + (d.cap ? E(d.cap) : "") + '">' +
        '<button class="primary" id="cu-cap-go" style="padding:7px 14px">Save</button></div>';
      h += '<div class="w-sub" id="cu-cap-msg" style="font-size:10.5px;min-height:14px"></div>';

      el.innerHTML = h;

      // wire the soft-cap control
      var inp = el.querySelector("#cu-cap"), go = el.querySelector("#cu-cap-go"), msg = el.querySelector("#cu-cap-msg");
      if (go) {
        go.onclick = function () {
          var v = (inp.value || "").trim();
          var cap = v === "" ? null : parseInt(v.replace(/[,_\s]/g, ""), 10);
          if (v !== "" && (isNaN(cap) || cap < 0)) {
            if (msg) msg.textContent = "Enter a whole number of tokens (or clear to remove).";
            return;
          }
          go.disabled = true;
          fetch("/api/claude_usage/config", {
            method: "POST", headers: { "Content-Type": "application/json" },
            body: JSON.stringify({ cap: cap })
          }).then(function (r) { return r.json(); }).then(function (x) {
            go.disabled = false;
            if (x && x.ok) {
              if (msg) msg.textContent = cap ? ("Soft cap set to " + cap.toLocaleString() + " tokens per 5-hour window.") : "Soft cap cleared.";
            } else if (msg) { msg.textContent = "Could not save."; }
          }).catch(function () { go.disabled = false; if (msg) msg.textContent = "Could not save."; });
        };
      }

      // subtle reveal on the daily bars (Motion One, fully guarded)
      try {
        if (typeof animate === "function") {
          var bars = el.querySelectorAll(".cu-bar");
          for (var i = 0; i < bars.length; i++) {
            animate(bars[i], { opacity: [0, 1], transform: ["scaleY(0.35)", "scaleY(1)"] },
              { duration: 0.5, delay: i * 0.035, easing: [0.2, 0.7, 0.3, 1] });
          }
        }
      } catch (e) {}
    };
  }
})();
